### PR TITLE
CPU: Parallel Kruskal MST with striped union-find and tests

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds CPU-based, parallel minimum spanning tree (MST) utilities using Kruskal's algorithm
- Introduces a striped, concurrent union-find to enable parallel cycle checks safely
- Exposes MST types and the `parallel_kruskal` function behind the `cpu` feature

## Changes
- New module: chutoro-core/src/mst/mod.rs
  - Parallel Kruskal implementation with canonical edge handling, weight checks, and deterministic tie-breaking
  - Public types: MstEdge, MinimumSpanningForest, MstError, MstErrorCode
  - Public function: parallel_kruskal
  - Note: internal helper `parallel_kruskal_from_edges` is implemented as `pub(crate)` and not part of the public API
- New module: chutoro-core/src/mst/union_find.rs
  - ConcurrentUnionFind with striped locking, safe parallel unions, and component tracking
- New tests: chutoro-core/src/mst/tests.rs
  - Unit tests for empty graphs, out-of-bounds nodes, non-finite weights, self-edges, simple connected graphs, disconnected graphs, and many equal-weight edges
- Public API exposure (cpu feature) in chutoro-core/src/lib.rs
  - Exposes MinimumSpanningForest, MstEdge, MstError, MstErrorCode, and `parallel_kruskal` under the `cpu` feature
- Documentation updates
  - docs/chutoro-design.md: notes on CPU parallel Kruskal design, including edge canonicalization, global sorting with Rayon, and striped-lock union-find
  - docs/roadmap.md: marks parallel Kruskal as completed in the roadmap

## Design & Implementation Notes
- Edges are interpreted as undirected and canonicalised to `(min(u, v), max(u, v))`; self-edges are ignored
- All weights are validated to be finite; non-finite weights error with a stable code
- Global edge sorting is performed with par_sort_unstable for parallelism, using `f32::total_cmp` for deterministic ordering
- Within each equal-weight bucket, cycle checks are performed in parallel using a striped-lock union-find to avoid deadlocks and improve throughput
- The union-find uses a striped lock table with a stable lock ordering to preserve correctness under concurrency

## Testing
- [x] Core behavior validated via unit tests (empty graph, invalid nodes, non-finite weights, self-edges, etc.)
- [x] Correctness checks for connected and disconnected graphs (forest vs. tree invariants)
- [x] Deterministic behavior for equal-weight edges with stable tie-breaks
- [x] Repeated stress-like scenarios (many equal weights) to ensure no cycles and correct MST size

## API Usage
- Enable the feature by building with CPU backend:
  - Cargo features: --features cpu
- Basic usage pattern:
  - Call parallel_kruskal(node_count, edges) to obtain a MinimumSpanningForest
  - Inspect edges() for the MST edges and component_count() to understand connectivity

## Test Plan
- [x] cargo test --features cpu
- [x] Validate all unit tests in chutoro-core/src/mst/tests.rs
- [x] Validate edge-cases in mst.rs/test suite

## Notes for Reviewers
- This change introduces a new CPU-only path and should be gated behind the `cpu` feature to avoid impacting non-CPU backends
- The test suite exercises both simple and edge-case graphs to ensure robust behavior under concurrency

📎 **Task**: https://www.terragonlabs.com/task/e6f47486-4d22-44a8-9e0f-c511fc9ccf5a

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[chutoro-core/src/mst/union_find.rs](https://github.com/leynos/chutoro/pull/66/files/c2ffe05dfd7ef837966906b02ee93e13f559d516#diff-8a8fcd980928bbee531143a9a8113c124b5e70497e94b9f8365c42d0d98d13e3)

Comment on lines +72 to +87

~~~rust
    fn execute_union_with_locks(
        &self,
        first_lock: usize,
        second_lock: usize,
        left: usize,
        right: usize,
    ) -> Result<UnionAttempt, MstError> {
        if first_lock == second_lock {
            let _guard = self.lock_stripe(first_lock)?;
            return self.try_union_after_lock(left, right);
        }

        let _first_guard = self.lock_stripe(first_lock)?;
        let _second_guard = self.lock_stripe(second_lock)?;
        self.try_union_after_lock(left, right)
    }
~~~

❌ New issue: [Excess Number of Function Arguments](https://codescene.io/projects/71838/delta?repo-id=1205554&review-id=66&biomarker=Excess+Number+of+Function+Arguments&filename=chutoro-core%2Fsrc%2Fmst%2Funion_find.rs&method=ConcurrentUnionFind.execute_union_with_locks)
ConcurrentUnionFind.execute_union_with_locks has 5 arguments, max arguments = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

## Excess Number of Function Arguments
chutoro-core/src/mst/union_find.rs: ConcurrentUnionFind.execute_union_with_locks

### What lead to degradation?
ConcurrentUnionFind.execute_union_with_locks has 5 arguments, max arguments = 4

### Why does this problem occur?
Functions with many arguments indicate either a) low cohesion where the function has too many responsibilities, or b) a missing abstraction that encapsulates those arguments.
The threshold for the Rust language is 4 function arguments.

### How to fix it?
Start by investigating the responsibilities of the function. Make sure it doesn't do too many things, in which case it should be split into smaller and more cohesive functions. Consider the refactoring [INTRODUCE PARAMETER OBJECT](https://refactoring.com/catalog/introduceParameterObject.html) to encapsulate arguments that refer to the same logical concept.

File: chutoro-core/src/mst/union_find.rs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+            if left_root == right_root {
+                return Ok(false);
+            }
+
+            let lock_pair = lock_order(self.lock_index(left_root), self.lock_index(right_root));
+
+            match self.execute_union_with_locks(lock_pair, left, right)? {
+                UnionAttempt::Done(result) => return Ok(result),
+                UnionAttempt::Retry => continue,
```
[the remark this answers is not in the document]
Comment (leynos) [reply chutoro-core/src/mst/union_find.rs]: @coderabbitai has this now been resolved?

Use sourcegraph analysis to determine your answer.

If the comment has been resolved, please mark it as resolved using the API.

**Comment by leynos** [comment]
@coderabbitai have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- The initial edge validation and canonicalisation in `prepare_edge_list` is fully sequential; if you expect large edge sets, consider making this stage parallel (e.g. using `par_iter().filter_map(...)`) to avoid a single-threaded bottleneck before the parallel sort.
- In `ConcurrentUnionFind::lock_stripe`, using `LockPoisoned` for an out-of-range lock index conflates two different failure modes; given `lock_index` should never produce an invalid index, it may be clearer to assert or use a distinct internal error path for this invariant violation and reserve `LockPoisoned` strictly for actual poisoning.

## Individual Comments

### Comment 1
<location> `chutoro-core/src/mst/tests.rs:9` </location>
<code_context>
+
+use super::{MstEdge, MstError, parallel_kruskal};
+
+fn harvest(edges: &[(usize, usize, f32, u64)]) -> EdgeHarvest {
+    EdgeHarvest::new(
+        edges
</code_context>

<issue_to_address>
**suggestion (testing):** Add tests for canonicalisation and deduplication of undirected edges

Current tests don’t directly assert that `(u, v)` and `(v, u)` map to the same canonical undirected edge or that duplicates are removed after sorting. Please add a focused test that builds an MST from something like `(0, 1, w, s1)` and `(1, 0, w, s2)` and checks that the result contains a single edge with `source == 0`, `target == 1`, and the expected deduplication behaviour.

Suggested implementation:

```rust
fn harvest(edges: &[(usize, usize, f32, u64)]) -> EdgeHarvest {
    EdgeHarvest::new(
        edges
            .iter()
            .map(|(source, target, weight, sequence)| {
                CandidateEdge::new(*source, *target, *weight, *sequence)
            })
            .collect(),
    )
}

#[test]
fn undirected_edges_are_canonicalised_and_deduplicated() -> Result<(), MstError> {
    // Two nodes (0 and 1) with two parallel undirected edges between them:
    // (0, 1, w, s1) and (1, 0, w, s2). These should canonicalise to the same
    // undirected edge and be deduplicated down to a single MST edge.
    let node_count = 2;
    let weight = 1.0_f32;
    let s1: u64 = 10;
    let s2: u64 = 20;

    let edges = [
        (0_usize, 1_usize, weight, s1),
        (1_usize, 0_usize, weight, s2),
    ];

    let harvest = harvest(&edges);
    // Build the MST from the candidate edges.
    let mst = parallel_kruskal(node_count, &harvest)?;

    // We expect a single undirected edge between 0 and 1.
    assert_eq!(mst.len(), 1);

    let edge = &mst[0];

    // Canonicalisation: the stored edge direction should be (0, 1).
    assert_eq!(edge.source, 0);
    assert_eq!(edge.target, 1);

    // The weight should match the input weight.
    assert_eq!(edge.weight, weight);

    // Deduplication behaviour: with equal weights, the lower sequence number
    // should win (or whatever the implementation defines as canonical).
    // This assertion assumes the lower sequence number is kept.
    assert_eq!(edge.sequence, s1);

    Ok(())
}

fn check_forest_invariants(node_count: usize, edges: &[MstEdge]) -> usize {

```

Depending on the existing definitions in `super` you may need to adjust:

1. `parallel_kruskal` call:
   - If the signature is `parallel_kruskal(node_count: usize, harvest: EdgeHarvest)`, drop the reference and pass `harvest` by value.
   - If it returns `Result<Vec<MstEdge>, MstError>` under a different alias, update the `mst` type accordingly.

2. `MstEdge` field names:
   - If fields are not named `source`, `target`, `weight`, and `sequence`, update the assertions to match the actual struct fields (e.g. `edge.u`, `edge.v`, `edge.seq`).

3. If `MstError` is not used as the error type for `parallel_kruskal`, change the test function return type from `Result<(), MstError>` to the appropriate error type or to `()` and handle errors explicitly with `unwrap()` / `expect()`.

4. If the deduplication rule for equal-weight edges is different (e.g. higher sequence wins, or first-seen wins), update the `assert_eq!(edge.sequence, s1);` line to match the intended behaviour, or adjust the test data so that the expected winner is unambiguous with respect to your implementation.
</issue_to_address>

### Comment 2
<location> `chutoro-core/src/mst/tests.rs:78-86` </location>
<code_context>
+    roots.len()
+}
+
+#[test]
+fn rejects_empty_graph() {
+    let result = parallel_kruskal(0, &EdgeHarvest::default());
+    assert!(matches!(result, Err(MstError::EmptyGraph)));
+}
+
+#[test]
+fn rejects_out_of_bounds_node_ids() {
+    let edges = harvest(&[(0, 3, 1.0, 0)]);
+    let result = parallel_kruskal(3, &edges);
+    assert!(matches!(
+        result,
+        Err(MstError::InvalidNodeId {
+            node: 3,
+            node_count: 3
+        })
+    ));
+}
+
+#[test]
+fn rejects_non_finite_weight() {
+    let edges = harvest(&[(0, 1, f32::NAN, 0)]);
+    let result = parallel_kruskal(2, &edges);
</code_context>

<issue_to_address>
**suggestion (testing):** Exercise all non-finite weight variants (NaN, +inf, -inf) explicitly

`rejects_non_finite_weight` currently only checks `f32::NAN`, but the implementation treats all non-finite values identically. Please extend this test (e.g., via parametrization or a small helper) to also cover `f32::INFINITY` and `f32::NEG_INFINITY`, asserting they all return `MstError::NonFiniteWeight` with the expected endpoints.

```suggestion
#[test]
fn rejects_non_finite_weight() {
    for &weight in &[f32::NAN, f32::INFINITY, f32::NEG_INFINITY] {
        let edges = harvest(&[(0, 1, weight, 0)]);
        let result = parallel_kruskal(2, &edges);
        assert!(
            matches!(
                result,
                Err(MstError::NonFiniteWeight { left: 0, right: 1 })
            ),
            "expected NonFiniteWeight error for weight={:?}",
            weight
        );
    }
}
```
</issue_to_address>

### Comment 3
<location> `chutoro-core/src/mst/tests.rs:88-98` </location>
<code_context>
+}
+
+#[test]
+fn ignores_self_edges() {
+    let edges = harvest(&[(0, 0, 1.0, 0), (0, 1, 2.0, 1)]);
+    let result = parallel_kruskal(2, &edges).expect("valid graph must succeed");
</code_context>

<issue_to_address>
**suggestion (testing):** Add a test for graphs with no usable edges but non-zero node_count

You already cover `EmptyGraph` and the case where self-edges are dropped when other valid edges exist. Please also add a test where `node_count > 0` but all edges are self-edges or otherwise filtered out (e.g. only `(0, 0, w, s)` and `(1, 1, w, s)` for `node_count = 2`), so that `edge_list` ends up empty and the MST result is a forest with `component_count == node_count` and `edges().is_empty()`.

```suggestion
#[test]
fn ignores_self_edges() {
    let edges = harvest(&[(0, 0, 1.0, 0), (0, 1, 2.0, 1)]);
    let result = parallel_kruskal(2, &edges).expect("valid graph must succeed");
    assert_eq!(result.component_count(), 1);
    assert_eq!(result.edges().len(), 1);
    assert_eq!(result.edges()[0].source(), 0);
    assert_eq!(result.edges()[0].target(), 1);
}

#[test]
fn no_usable_edges_yields_forest() {
    let node_count = 2;
    let edges = harvest(&[(0, 0, 1.0, 0), (1, 1, 2.0, 1)]);
    let result =
        parallel_kruskal(node_count, &edges).expect("graph with only self-edges must succeed");
    assert_eq!(result.component_count(), node_count);
    assert!(result.edges().is_empty());
}

fn connected_unique_weights_case() -> (usize, EdgeHarvest, Vec<MstEdge>) {
```
</issue_to_address>

### Comment 4
<location> `chutoro-core/src/mst/tests.rs:98` </location>
<code_context>
+    assert_eq!(result.edges()[0].target(), 1);
+}
+
+fn connected_unique_weights_case() -> (usize, EdgeHarvest, Vec<MstEdge>) {
+    let node_count = 4;
+    let edges = harvest(&[
</code_context>

<issue_to_address>
**suggestion (testing):** Add tests for MST determinism and tie-breaking, including sequence-based ordering

`connected_unique_weights_case` only covers strictly increasing weights, so it doesn’t exercise the tie-breaking logic. Since the implementation sorts by `weight`, then `(source, target)`, then `sequence`, please add a test with multiple edges sharing the same weight but differing in `sequence` (and ideally with ambiguous `source/target` before canonicalisation). That test should assert the precise `MstEdge` order, including `.sequence()`, to lock in deterministic tie-breaking behavior.

Suggested implementation:

```rust
fn connected_unique_weights_case() -> (usize, EdgeHarvest, Vec<MstEdge>) {
    let node_count = 4;
    let edges = harvest(&[
        (0, 1, 1.0, 0),
        (1, 2, 2.0, 1),
        (2, 3, 3.0, 2),
        (0, 2, 6.0, 3),
        (0, 3, 10.0, 4),
    ]);

    let expected = vec![
        MstEdge {
            source: 0,
            target: 1,
            weight: 1.0,
            sequence: 0,
        },
        MstEdge {
            source: 1,
            target: 2,
            weight: 2.0,
            sequence: 1,
        },
        MstEdge {
            source: 2,
            target: 3,
            weight: 3.0,
            sequence: 2,
        },
    ];

    (node_count, edges, expected)
}

fn connected_tied_weights_case() -> (usize, EdgeHarvest, Vec<MstEdge>) {
    // All edges have the same weight, so ordering is determined by
    // (source, target) after canonicalisation, then by `sequence`.
    //
    // We intentionally include both (0, 2) and (2, 0) so that canonicalisation
    // to (0, 2) is required, and the `sequence` field is the final
    // tie-breaker between those two otherwise-identical edges.
    let node_count = 3;
    let edges = harvest(&[
        // Distinct endpoints
        (0, 1, 1.0, 7),
        (1, 2, 1.0, 8),
        // Same endpoints after canonicalisation, different sequences
        (2, 0, 1.0, 5),
        (0, 2, 1.0, 10),
    ]);

    let expected = vec![
        // Expected deterministic MST order:
        // 1. (0, 1) — lowest (source, target) among weight ties
        // 2. (0, 2) — chosen with the lowest sequence among edges between 0 and 2
        MstEdge {
            source: 0,
            target: 1,
            weight: 1.0,
            sequence: 7,
        },
        MstEdge {
            source: 0,
            target: 2,
            weight: 1.0,
            sequence: 5,
        },
    ];

    (node_count, edges, expected)
}

#[test]
fn mst_is_deterministic_with_tied_weights_and_sequences() {
    let (node_count, edges, expected) = connected_tied_weights_case();
    let result = parallel_kruskal(node_count, &edges).expect("valid graph must succeed");

    // Single connected component with a fully-determined, stable edge order.
    assert_eq!(result.component_count(), 1);
    assert_eq!(result.edges().len(), expected.len());

    // Assert exact MstEdge ordering, including `.sequence()`.
    for (actual, expected_edge) in result.edges().iter().zip(expected.iter()) {
        assert_eq!(actual.source(), expected_edge.source);
        assert_eq!(actual.target(), expected_edge.target);
        assert_eq!(actual.weight(), expected_edge.weight);
        assert_eq!(actual.sequence(), expected_edge.sequence);
    }

```

If the `MstEdge` struct uses different field names or types (e.g. `seq` instead of `sequence`, or a non-`f64` weight), adjust the struct literals and the `expected_edge` field accesses accordingly.

If `parallel_kruskal` returns edges in a different order than the sort order described (e.g. if it performs a final re-sort before returning), update the `expected` vector in `connected_tied_weights_case` to match the actual, intended deterministic ordering while keeping the same tied-weight/tied-sequence scenario.

If `MstEdge` does not have the accessor methods `source()`, `target()`, `weight()`, or `sequence()`, replace those calls in the test with direct field accesses matching your API.
</issue_to_address>

### Comment 5
<location> `chutoro-core/src/mst/tests.rs:134` </location>
<code_context>
+
+#[rstest]
+#[case::connected_unique_weights(connected_unique_weights_case())]
+fn returns_expected_mst_on_unique_weights(#[case] case: (usize, EdgeHarvest, Vec<MstEdge>)) {
+    let (node_count, edges, expected) = case;
+    let result = parallel_kruskal(node_count, &edges).expect("MST must succeed");
</code_context>

<issue_to_address>
**suggestion (testing):** Assert `is_tree()` and the `node_count - 1` edge invariant for connected graphs

This test already checks `component_count == 1` and the exact edge set. To better cover the API invariants, please also assert `result.is_tree()` and `result.edges().len() == node_count - 1` here (and in at least one other connected-graph test) to exercise the `MinimumSpanningForest` helpers and the spanning-tree edge-count property.

Suggested implementation:

```rust
#[rstest]
#[case::connected_unique_weights(connected_unique_weights_case())]
fn returns_expected_mst_on_unique_weights(#[case] case: (usize, EdgeHarvest, Vec<MstEdge>)) {
    let (node_count, edges, expected) = case;
    let result = parallel_kruskal(node_count, &edges).expect("MST must succeed");

    assert_eq!(result.component_count(), 1);
    assert!(result.is_tree(), "MST of a connected graph should be a single tree");
    assert_eq!(
        result.edges().len(),
        node_count - 1,
        "MST of a connected graph should have N-1 edges"
    );
    assert_eq!(result.edges(), expected.as_slice());
}

```

To fully implement your suggestion:
1. Identify at least one other test that exercises a connected graph and receives a `MinimumSpanningForest` (or equivalent) result.
2. In that test, after asserting connectivity or expected edges, add:
   - `assert!(result.is_tree());`
   - `assert_eq!(result.edges().len(), node_count - 1);`
   mirroring the pattern introduced in `returns_expected_mst_on_unique_weights`.
3. Ensure that `node_count` (or an equivalent variable) is in scope for the additional test so the `node_count - 1` invariant can be asserted.
</issue_to_address>

### Comment 6
<location> `chutoro-core/src/mst/tests.rs:163` </location>
<code_context>
+}
+
+#[test]
+fn handles_many_equal_weights_without_cycles() {
+    let node_count = 6;
+    let edges = harvest(&[
</code_context>

<issue_to_address>
**suggestion (testing):** Strengthen the equal-weight stress test to also check deterministic ordering

The current loop in `handles_many_equal_weights_without_cycles` does a good job validating structural properties (component count, edge count, uniform weight), but it doesn’t assert determinism. To also cover the deterministic ordering guarantee, consider capturing the MST edge list (including `sequence()`) from the first iteration and asserting that each subsequent iteration produces the exact same ordered list. This would more directly verify that the parallel algorithm and striped union-find yield a stable, deterministic forest across runs.

Suggested implementation:

```rust
#[test]
fn handles_many_equal_weights_without_cycles() {
    let node_count = 6;
    let edges = harvest(&[
        // existing equal-weight, acyclic edge definitions stay as-is
    ]);

    // Run the MST algorithm multiple times to stress determinism.
    // Capture the ordered edge list (including `sequence()`) from the first run
    // and assert that all subsequent runs produce the exact same ordered list.
    let mut first_run_edges: Option<Vec<MstEdge>> = None;

    for _ in 0..32 {
        let result = parallel_mst(node_count, edges.clone());

        // Preserve the existing structural checks.
        let component_count = check_forest_invariants(node_count, result.edges());
        assert_eq!(result.component_count(), component_count);
        assert_eq!(
            result.edges().len(),
            node_count.saturating_sub(component_count)
        );

        // Capture the ordered edge list as produced by the algorithm.
        let current_edges: Vec<MstEdge> = result.edges().cloned().collect();

        match &mut first_run_edges {
            None => {
                // First iteration: record the canonical ordered forest.
                first_run_edges = Some(current_edges);
            }
            Some(expected_edges) => {
                // Subsequent iterations must produce the exact same ordered forest.
                assert_eq!(&current_edges, expected_edges);
            }
        }
    }

```

1. Replace `parallel_mst(node_count, edges.clone())` with the actual MST entry point used elsewhere in this test module (e.g. `build_parallel_mst(node_count, &edges)` or similar), matching its existing call signature.
2. Ensure `edges` is clonable or otherwise reusable across loop iterations (e.g. if `harvest` returns `Vec<_>`, cloning is fine; if it returns `Arc<[T]>` or similar, adjust accordingly).
3. Confirm that `MstEdge` implements `Clone` and `PartialEq` (or `Eq`); if it doesn’t, derive or implement them so `cloned()` and `assert_eq!` compile.
4. Remove or adjust any pre-existing loop or assertions inside `handles_many_equal_weights_without_cycles` that duplicate the structural checks, so the function body matches this new structure exactly.
</issue_to_address>

### Comment 7
<location> `chutoro-core/src/mst/union_find.rs:14` </location>
<code_context>
+
+use super::MstError;
+
+pub(super) struct ConcurrentUnionFind {
+    parents: Vec<AtomicUsize>,
+    ranks: Vec<AtomicUsize>,
</code_context>

<issue_to_address>
**issue (complexity):** Consider simplifying the union-find concurrency design by replacing striped locks and the retry state machine with per-node ordered locking and a single-pass union flow.

You can meaningfully reduce complexity without changing behaviour by simplifying the locking model and removing the retry state machine.

### 1. Replace striped locks with per-node locks

You already lock based on roots; using a stripe mask adds an extra layer (`lock_index`, `lock_stripes`, `lock_mask`) that isn’t strictly needed for parallel Kruskal.

You can keep parallelism and deadlock-free ordering by using one `Mutex<()>` per node and always locking `(min_root, max_root)`:

```rust
pub(super) struct ConcurrentUnionFind {
    parents: Vec<AtomicUsize>,
    ranks: Vec<AtomicUsize>,
    components: AtomicUsize,
    // one lock per node instead of stripes + mask
    locks: Vec<Mutex<()>>,
}

impl ConcurrentUnionFind {
    pub(super) fn new(node_count: usize) -> Self {
        let mut parents = Vec::with_capacity(node_count);
        let mut ranks = Vec::with_capacity(node_count);
        for id in 0..node_count {
            parents.push(AtomicUsize::new(id));
            ranks.push(AtomicUsize::new(0));
        }

        let locks = (0..node_count).map(|_| Mutex::new(())).collect();

        Self {
            parents,
            ranks,
            components: AtomicUsize::new(node_count),
            locks,
        }
    }

    fn lock_two(&self, a: usize, b: usize) -> Result<(MutexGuard<'_, ()>, MutexGuard<'_, ()>), MstError> {
        let (first, second) = if a <= b { (a, b) } else { (b, a) };
        let first_guard = self.locks.get(first)
            .ok_or(MstError::LockPoisoned { resource: "union-find lock" })?
            .lock()
            .map_err(|_| MstError::LockPoisoned { resource: "union-find lock" })?;
        if first == second {
            return Ok((first_guard, first_guard));
        }
        let second_guard = self.locks.get(second)
            .ok_or(MstError::LockPoisoned { resource: "union-find lock" })?
            .lock()
            .map_err(|_| MstError::LockPoisoned { resource: "union-find lock" })?;
        Ok((first_guard, second_guard))
    }
}
```

This lets you drop `lock_mask`, `lock_index`, `lock_stripes`, and `lock_order` entirely.

### 2. Remove `UnionAttempt` and the retry loop

You can fold `execute_union_with_locks` + `UnionAttempt` into a single, simpler `try_union` that does at most one retry-equivalent by re-reading the roots under the locks and returning `Ok(false)` if they changed.

Conceptually:

1. Compute `left_root` / `right_root`.
2. If equal, return `Ok(false)`.
3. Lock `(min(left_root, right_root), max(left_root, right_root))`.
4. Recompute roots under the locks.
5. If they changed or are equal now, return `Ok(false)`.
6. Perform `union_roots`.

Example:

```rust
pub(super) fn try_union(&self, left: usize, right: usize) -> Result<bool, MstError> {
    let mut left_root = self.find(left);
    let mut right_root = self.find(right);

    if left_root == right_root {
        return Ok(false);
    }

    // Lock based on root ids to keep order consistent
    let (_g1, _g2) = self.lock_two(left_root, right_root)?;

    // Recompute under the locks to detect races
    left_root = self.find(left);
    right_root = self.find(right);

    if left_root == right_root {
        return Ok(false);
    }

    // If either changed to non-root, treat as "no union" to avoid the state machine
    if !self.is_root(left_root) || !self.is_root(right_root) {
        return Ok(false);
    }

    self.union_roots(left_root, right_root)
}
```

With this, you can remove:

- `UnionAttempt` enum,
- `execute_union_with_locks`,
- `try_union_after_lock`,
- the outer `loop` in `try_union`.

The net effect is:

- same observable behaviour from callers (`Result<bool, MstError>`),
- no deadlocks (consistent `(min, max)` locking),
- parallelism preserved (per-node locks allow disjoint unions to proceed in parallel),
- much simpler mental model: “find → lock two roots in order → re-validate → union or no-op.”
</issue_to_address>

### Comment 8
<location> `chutoro-core/src/mst/mod.rs:316` </location>
<code_context>
+/// CPU minimum spanning tree (MST) utilities; requires the `cpu` feature.
+pub use crate::mst::{MinimumSpanningForest, MstEdge, MstError, MstErrorCode, parallel_kruskal};
+
 #[cfg(test)]
 pub(crate) mod test_utils;
--- a/docs/roadmap.md
</code_context>

<issue_to_address>
**issue (review_instructions):** Add both behavioural and unit tests for the new parallel Kruskal MST implementation instead of leaving the test module effectively empty.

You introduce a substantial new feature here (the parallel Kruskal MST implementation and `ConcurrentUnionFind`), but there are no accompanying tests in `mst/tests.rs`. Implement unit tests that directly exercise `parallel_kruskal_from_edges`, the error conditions (`EmptyGraph`, `InvalidNodeId`, `NonFiniteWeight`, `LockPoisoned`), and the union‑find behaviour, and add higher‑level behavioural tests that validate MST/forest correctness for representative graphs (connected, disconnected, equal‑weight edges, self‑loops, duplicate edges, etc.). Ensure these tests are actually implemented in `mst/tests.rs` so this module satisfies the requirement for both behavioural and unit tests for new features.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*`

**Instructions:**
Rules:
- For any new feature or change to an existing feature, both behavioural *and* unit tests are required.
- Bug fixes *must* be demonstrated by a test.

</details>
</issue_to_address>

### Comment 9
<location> `chutoro-core/src/mst/union_find.rs:115` </location>
<code_context>
+        Ok(UnionAttempt::Done(self.union_roots(left_root, right_root)?))
+    }
+
+    fn union_roots(&self, left_root: usize, right_root: usize) -> Result<bool, MstError> {
+        let left_rank = self.ranks[left_root].load(Ordering::Relaxed);
+        let right_rank = self.ranks[right_root].load(Ordering::Relaxed);
</code_context>

<issue_to_address>
**issue (review_instructions):** The `union_roots` function contains an if/else-if/else-if/else chain with more than two branches, which should be extracted into a separate predicate/helper function.

Within `union_roots`, the parent/child selection logic uses a four-branch conditional chain:

```rust
if left_rank > right_rank {
    ...
} else if right_rank > left_rank {
    ...
} else if left_root <= right_root {
    ...
} else {
    ...
}
```

Per the review instructions, conditionals with more than two branches should be moved into a predicate (or helper) function. For example, you could extract a helper like `fn choose_parent_child(left_root, right_root, left_rank, right_rank) -> (usize, usize)` and call it from `union_roots`.

That would keep `union_roots` focused on performing the union and maintain the requested style for multi-branch conditionals.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Move conditionals with >2 branches into a predicate function.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[chutoro-core/src/mst/tests.rs](https://github.com/leynos/chutoro/pull/66/files/348fcba559b9d985138cf201379e1709325417c8#diff-4c5cea8d5b48dd9f98344e4838c54e3e75f10d9c33f085c3f06b099347f89c25)

Comment on lines +153 to +185

~~~rust
fn connected_unique_weights_case() -> (usize, EdgeHarvest, Vec<MstEdge>) {
    let node_count = 4;
    let edges = harvest(&[
        (0, 1, 1.0, 0),
        (1, 2, 2.0, 1),
        (2, 3, 3.0, 2),
        (0, 2, 6.0, 3),
        (0, 3, 10.0, 4),
    ]);

    let expected = vec![
        MstEdge {
            source: 0,
            target: 1,
            weight: 1.0,
            sequence: 0,
        },
        MstEdge {
            source: 1,
            target: 2,
            weight: 2.0,
            sequence: 1,
        },
        MstEdge {
            source: 2,
            target: 3,
            weight: 3.0,
            sequence: 2,
        },
    ];

    (node_count, edges, expected)
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71838/delta?repo-id=1205554&review-id=66&biomarker=Code+Duplication&filename=chutoro-core%2Fsrc%2Fmst%2Ftests.rs&method=)
The module contains 2 functions with similar structure: connected_duplicate_edges_case,connected_unique_weights_case
